# interproscan6

**!! UNDER DEVELOPMENT !!**

## Requirements

* `Java` (version >= 11)
* `Nextflow` (version >=23.04.02)

## Set up

1. Download data files:
- members bin (you can find this in interproscan5 on path `interproscan/core/jms-implementation/target/interproscan-5-dist/bin`)
- members data (`curl ftp://ftp.ebi.ac.uk/pub/software/unix/iprscan/5/5.62-94.0/alt/interproscan-data-5.62-94.0.tar.gz --output interproscan-data-5.62-94.0.tar.gz`)
- xref (entries, goterms and pathways) files (Use this script: `interproscan6/files_test/get_data_to_i6.py`)

2. Build a docker `interproscan6` image (this is necessary for the HMMer process):

    docker build -t interproscan6 .
    
3. Download the additional images we use:
    
    docker pull staphb/hmmer

## Quick start

How to run:


    nextflow run interproscan.nf --input <path to fasta file> [resume]


The results will apear in `result/` folder.  
For debugging, you an find all working files generated by the pipeline in the `work/` dir.

Batchsize parameter in `nextflow.config` defines the number maximum number of sequences submitted within each batch, and thus the quantity of threads of parallelism.

**IMPORTANT:** Change the input params in the `input.yaml` file if you want to test different flows (see in `main.nf`)

## Inputs

### Configuration

#### Parameters

Parameters can be set on command line using double dashes. The only mandatory parameter is the --input one.

`IMPORTANT: Parameters that have more than one value must be separated by comma and without space (e.g. --applications antifam,ncbifam,pfam)`

**Configuration parameters:**

* `--applications` - a list of member databases/applications to be inclued in the analysis. For example:

```yaml
applications: AntiFam,CDD,Coils,FunFam,Gene3d,HAMAP,MobiDBLite,NCBIfam,Panther,Pfam,Phobius,PIRSF,PIRSR,PRINTS,PrositePatterns,PrositeProfiles,SFLD,SignalP_EUK,SignalP_GRAM_NEGATIVE,SignalP_GRAM_POSITIVE,SMART,SuperFamily,TMHMM
```

* `--disable_precalc` - Do not run comparison against an InterPro release to retrive precalculated matches, instead run `interproscan` for all input sequences. [Boolean]
* `--formats` - List output file formats. Supported: JSON, TSV, GFF, XML
* `--goterms` - Whether to retrieve and include Gene Ontology terms from InterPro in the output files. [Boolean]
* `--help` - Whether to disble the help message - `interproscan-6` will not run any analysis when `help` is set to true. [Boolean]
* `--input` - Path to input FASTA file
* `--outfile` - Path and prefix for output files
* `--pathways` - Optional, switch on lookup of corresponding Pathway annotation (IMPLIES - `lookup_file` is defined) [Boolean]
* `--lookup_file` - Lookup of corresponding InterPro annotation in the TSV and GFF3 output formats.

#### Nextflow configuration

The batch sizes, and paths to the InterPro release datasets are defined in the `./nextflow.config` file. 

### Sequences

The input sequences to be analysed by `interproscan-6`must be provided in FASTA format. All input sequences must be provided in a multi-sequence FASTA file.

At the moment only protein (amino acid) sequences are supported.


## Outputs

:TODO:
