# InterProScan6

**!! UNDER DEVELOPMENT !!**

## Requirements

* `Java` (version >= 11)
* `Nextflow` (version >=23.04.02)

## Set up

1. Download member data files:
```bash
curl ftp://ftp.ebi.ac.uk/pub/software/unix/iprscan/5/5.67-99.0/alt/interproscan-data-5.67-99.0.tar.gz \
    --output interproscan-data-5.67-99.0.tar.gz
tar -pxzf interproscan-data-5.67-99.0.tar.gz
mv interproscan-5.67-99.0/data .
rm interproscan-5.67-99.0 -rf
rm interproscan-data-5.67-99.0.tar.gz
```

2. Download InterPro data files (xref (entries, goterms and pathways)):
```bash
mkdir i6data
python interproscan6/files_test/get_data_to_i6.py  # for devs
```

3. Build a docker `InterProScan6` base image:

    docker build -t interproscan6 .
    
4. Download the additional images we use:
    
    docker pull biocontainers/hmmer:v3.2.1dfsg-1-deb_cv1

5. [Optional] install licensed software

By default `Phobius`, `SignalP`, and `TMHMM` member database analyses are deactivated in `InterProScan6` 
because they contain licensed components. In order to activate these analyses please see the ['Installing licensed applications'](#installing-licensed-applications-phobius-signalp-tmhmm) documentation.

## Quick start

How to run:

    nextflow run interproscan.nf --input <path to fasta file> [resume]

The results will apear in `result/` folder.  
For debugging, you an find all working files generated by the pipeline in the `work/` dir.

Batchsize parameter in `nextflow.config` defines the number maximum number of sequences submitted within each batch, and thus the quantity of threads of parallelism.

**IMPORTANT:** Change the input params in the `input.yaml` file if you want to test different flows (see in `main.nf`)

## Inputs

### Configuration

#### Parameters

`InterProScan6` is configured via the command-line. The only mandatory parameter is `--input`.

**IMPORTANT:** For parameters that have more than one value, separate values using a comma (not spaces) (e.g. `--applications antifam,ncbifam,pfam)`)

**Configuration parameters:**

* `--applications` - a list of member databases/applications to be employed in the analysis. By default, `InterProScan6` employs all member databases and applications, use this flag to run only a subset of applications. For example:
```bash
nextflow run interproscan.nf --input <fasta file> --applications NCBIfam,Panther,Pfam

```yaml
applications: AntiFam,CDD,Coils,FunFam,Gene3d,HAMAP,MobiDBLite,NCBIfam,Panther,Pfam,Phobius,PIRSF,PIRSR,PRINTS,PrositePatterns,PrositeProfiles,SFLD,SignalP_EUK,SignalP_GRAM_NEGATIVE,SignalP_GRAM_POSITIVE,SMART,SuperFamily,TMHMM
```

* `--disable_precalc` - Do not run comparison against an InterPro release to retrive precalculated matches, instead run `interproscan` for all input sequences. [Boolean]
* `--formats` - List output file formats. Supported: JSON,TSV,GFF,XML
* `--goterms` - Whether to retrieve and include Gene Ontology terms from InterPro in the output files. [Boolean]
* `--help` - Whether to disble the help message - `interproscan-6` will not run any analysis when `help` is set to true. [Boolean]
* `--input` - Path to input FASTA file
* `--outfile` - Path and prefix for output files
* `--pathways` - Optional, switch on lookup of corresponding Pathway annotation (IMPLIES - `lookup_file` is defined) [Boolean]
* `--lookup_file` - Lookup of corresponding InterPro annotation in the TSV and GFF3 output formats.

#### Nextflow configuration

The batch sizes, and paths to the InterPro release datasets are defined in the `./nextflow.config` file. 

### Sequences

The input sequences to be analysed by `interproscan-6`must be provided in FASTA format. All input sequences must be provided in a multi-sequence FASTA file.

At the moment only protein (amino acid) sequences are supported.


## Outputs

:TODO:

# Installing licensed applications (`Phobius`, `SignalP`, `TMHMM`)

By default `Phobius`, `SignalP`, and `TMHMM` member database analyses are deactivated in `InterProScan6` 
because they contain licensed components. In order to activate these analyses please 
obtain the relevant license and files from the provider (ensuring the software version 
numbers are the same as those supported by your current `InterProScan6` installation).

Files can be placed in any location as long as your `subworkflows/sequence_analysis/members.config` configuration is updated accordingly.

## `SignalP`

### Adding `SignalP` (version 6) to `InterProScan6`

1. Follow the installation instructions for `SignalP` (version 6) available [here](https://github.com/fteufel/signalp-6.0/blob/main/installation_instructions.md)
    * Either fast or slow models can be implemented
    * To change the implemented mode please see the [Changing mode](#changing-mode) documentation

2. Update the `InterProScan6` configuration: update `subworkflows/sequence_analysis/members.config`:
```
signalp {
    runner = "signalp"
    mode = "fast"   <--- UPDATE MODE
    data = {
        model_dir = "$projectDir/bin/signalp/models"  <--- UPDATE PATH TO models DIR
    }
    switches = {
        other = "--organism other -f summary -c 70"
        euk = "--organism euk -f summary -c 70"
    }
}
```
Note: _Instead of updating the `model_dir` parameter, you could copy the models directory contents into the `InterProScan6/signalp/models` directory._

### Running `InterProScan6` with `SignalP6` enabled

Include `SignalP_OTHER` (to see all models in `SignalP6`) or `SignalP_EUK` (to limit predictions to Sec/SPI) to the `applications` parameter in `nextflow.config`

```
params {
    batchsize = 100
    help = false
    applications = 'SignalP_OTHER'  <---
    disable_precalc = false
    tsv_pro = false
}
```

Specifying the eukarya method of `SignalP6` (`SignalP_EUK`) triggers post-processing of the SP predictions by `SignalP6` to prevent spurious results (only predicts type Sec/SPI).

### Changing mode

`SignalP6` supports 3 modes: `fast`, `slow` and `slow-sequential`. To change the mode of `SignalP6`:

1. Incorporate the new mode into your `SignalP6` installtion, please follow the `SignalP6` [documentation](https://github.com/fteufel/signalp-6.0/blob/main/installation_instructions.md#installing-additional-modes).

2. Update the `member.config` configuration (`subworkflows/sequence_analysis/members.config`)
```
signalp {
    runner = "signalp"
    mode = "fast"   <--- UPDATE MODE
    data = {
        model_dir = "$projectDir/bin/signalp/models"
    }
    switches = {
        other = "--organism other -f summary -c 70"
        euk = "--organism euk -f summary -c 70"
    }
}
```

### Converting from CPU to GPU, and back again

:TODO: -- add support for GPU run

The model weights that come with the `SignalP` installation by default run on your CPU.
If you have a GPU available, you can convert your installation to use the GPU instead. 

1. Convert the `SignalP` installation to GPU by following the `SignalP` [documentation](https://github.com/fteufel/signalp-6.0/blob/main/installation_instructions.md#converting-to-gpu)
2. ....???....

# Trouble shooting

## Permission denied

On some systems, Nextflow requires root privileges to be able to create the output directories. 

If you receive an error message such as:
```bash
ERROR ~ Error executing process > 'PARSE_SEQUENCE (1)'

Caused by:
  Unable to create directory=.../InterProScan6/work/56/c74d6bbc6637c201a68afd84b75d27 -- check file system permissions
```

Try running Nextflow with root privileges:
```
sudo nextflow run interproscan.nf --input <path to fasta file> 
```

## File not found

If you recieve a file not found error:
```bash
FileNotFoundError: [Errno 2] No such file or directory
```

This may be due to the docker image not being built correctly. This can happen due to file permissions from preventing docker from reading the `InterProScan6` files.

Try running docker with root privileges:
```
sudo docker build -t interproscan6 .
```

Check the docker installtion is configured correctly, with all necessary privileges. [StackOverflow](https://stackoverflow.com/questions/48957195/how-to-fix-docker-got-permission-denied-issue)
